File: all/3546_equal_sum_grid_partition_I/sol_better_space.c
#include <stdbool.h>
#include <stdlib.h>

bool canPartitionGrid(int** grid, int gridSize, int* gridColSize) {
    if (0 == gridSize) {
        return false;
    }

    const int GRID_MATRIX_ROWS_NUMBER = gridSize;
    const int GRID_MATRIX_COLUMNS_NUMBER = (gridColSize[0]);

    bool result = false;
    long long total_sum = 0, accumulated_sum = 0;
    long long *data_blocks_array_ptr = (
        (long long*)calloc(
            GRID_MATRIX_ROWS_NUMBER + GRID_MATRIX_COLUMNS_NUMBER,
            sizeof(*data_blocks_array_ptr)
        )
    );
    long long *row_sum_array_ptr = data_blocks_array_ptr;
    long long *column_sum_array_ptr = (
        data_blocks_array_ptr + GRID_MATRIX_ROWS_NUMBER
    );

    for (int row_index = 0; row_index < GRID_MATRIX_ROWS_NUMBER; ++row_index) {
        for (
            int column_index = 0;
            column_index < GRID_MATRIX_COLUMNS_NUMBER;
            ++column_index
        ) {
            row_sum_array_ptr[row_index] += (grid[row_index][column_index]);
            column_sum_array_ptr[column_index] += (
                grid[row_index][column_index]
            );
            total_sum += (grid[row_index][column_index]);
        }
    }

    if (total_sum & 0x1) {
        /*
         * If total sum is an odd number, it is impossible that we can cut
         * the matrix into two parts with equal sum.
         */
        result = false;

        goto FINAL_CLEANUP;
    }

    accumulated_sum = 0;

    for (
        int column_index = 0;
        column_index < (GRID_MATRIX_COLUMNS_NUMBER - 1);
        ++column_index
    ) {
        accumulated_sum += (column_sum_array_ptr[column_index]);

        if (total_sum == (accumulated_sum << 1)) {
            result = true;

            goto FINAL_CLEANUP;
        }
    }

    accumulated_sum = 0;

    for (
        int row_index = 0;
        row_index < (GRID_MATRIX_ROWS_NUMBER - 1);
        ++row_index
    ) {
        accumulated_sum += (row_sum_array_ptr[row_index]);

        if (total_sum == (accumulated_sum << 1)) {
            result = true;

            goto FINAL_CLEANUP;
        }
    }

FINAL_CLEANUP:
    free(data_blocks_array_ptr);
    data_blocks_array_ptr = NULL;
    row_sum_array_ptr = NULL;
    column_sum_array_ptr = NULL;

    return result;
}

File: all/3546_equal_sum_grid_partition_I/sol_better_space.py
from typing import List


class Solution:
    def canPartitionGrid(self, grid: List[List[int]]) -> bool:
        if not grid:
            return False

        GRID_MATRIX_ROWS_NUMBER = len(grid)
        GRID_MATRIX_COLUMNS_NUMBER = len(grid[0])

        row_sum_list = [0] * GRID_MATRIX_ROWS_NUMBER
        column_sum_list = [0] * GRID_MATRIX_COLUMNS_NUMBER
        total_sum = 0

        for row_index, row in enumerate(grid):
            for column_index, current_cell_value in enumerate(row):
                row_sum_list[row_index] += current_cell_value
                column_sum_list[column_index] += current_cell_value
                total_sum += current_cell_value

        # Try every possible horizontal cut.
        accumulated_sum = 0

        for row_index in range(GRID_MATRIX_ROWS_NUMBER - 1):
            accumulated_sum += row_sum_list[row_index]

            if total_sum == (accumulated_sum << 1):
                return True

        # Try every possible vertical cut.
        accumulated_sum = 0

        for column_index in range(GRID_MATRIX_COLUMNS_NUMBER - 1):
            accumulated_sum += column_sum_list[column_index]

            if total_sum == (accumulated_sum << 1):
                return True

        return False


if '__main__' == __name__:
    sol = Solution()
    testcases_list = [
        [[[1, 4], [2, 3]]],
        [[[1, 3], [2, 4]]],
    ]

    for testcase in testcases_list:
        print(sol.canPartitionGrid(*testcase))
